src/rumor_state.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// http://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+use crate::id::Id;
+use std::collections::btree_map::Entry;
+use std::collections::{BTreeMap, BTreeSet};
+
+/// This represents the state of a single rumor from this node's perspective.
+#[derive(Debug)]
+pub enum RumorState {
+    /// Exponential-growth phase.
+    B {
+        /// The round number for this rumor.  This is not a globally-synchronised variable, rather
+        /// it is set to 0 when we first receive a copy of this rumor and is incremented every
+        /// time `next_round()` is called.
+        round: u8,
+        /// Our age for this rumor.  This may increase by 1 during a single round or may
+        /// remain the same depending on the ages attached to incoming copies of this rumor.
+        our_age: u8,
+        /// The map of <peer, age>s which have sent us this rumor during this round.
+        peer_ages: BTreeMap<Id, u8>,
+    },
+    /// Quadratic-shrinking phase.
+    C {
+        /// The number of rounds performed by the node while the rumor was in state B.
+        rounds_in_state_b: u8,
+        /// The round number for this rumor while in state C.
+        round: u8,
+    },
+    /// Propagation complete.
+    D,
+}
+
+impl RumorState {
+    /// Construct a new `RumorState` where we're the initial node for the rumor.  We start in
+    /// state B with `our_age` set to `1`.
+    pub fn new() -> Self {
+        RumorState::B {
+            round: 0,
+            our_age: 1,
+            peer_ages: BTreeMap::new(),
+        }
+    }
+
+    /// Construct a new `RumorState` where we've received the rumor from a peer.  If that peer
+    /// is in state B (`age < max_b_age`) we start in state B with `our_age` set to `1`.
+    /// If the peer is in state C, we start in state C too.
+    pub fn new_from_peer(age: u8, max_b_age: u8) -> Self {
```

Comment:
It might be good to set a type `RomourAge` to ensure we don't confuse with node age. Also sometimes it helps to read code instead of a bare u8?